pkg/kubectl/cmd/set/set_data.go

```diff
@@ -0,0 +1,375 @@
+		FilenameParam(enforceNamespace, &o.fileNameOptions).
+		IncludeUninitialized(includeUninitialized).
+		Flatten()
+	if !o.Local {
+		builder.ResourceTypeOrNameArgs(o.all, resources...).
+			Latest()
+	}
+	o.Infos, err = builder.Do().Infos()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (o *DataOptions) Validate() error {
+	errors := []error{}
+	if len(o.EnvFileSource) > 0 && (len(o.FileSources) > 0 || len(o.LiteralSources) > 0) {
+		errors = append(errors, fmt.Errorf("from-env-file cannot be combined with from-file or from-literal"))
+	}
+	return utilerrors.NewAggregate(errors)
+}
+
+// Run creates and applies the patch either Locally or calling apiserver.
+func (o *DataOptions) Run() error {
+	patches := CalculatePatches(o.Infos, cmdutil.InternalVersionJSONEncoder(), func(info *resource.Info) ([]byte, error) {
+		transformed, err := updateDataForObject(info.Object, o.Data, addData)
+		if transformed && err == nil {
+			return runtime.Encode(cmdutil.InternalVersionJSONEncoder(), info.AsVersioned())
+		}
+		return nil, err
+	})
+
+	allErrs := []error{}
+	for _, patch := range patches {
+		info := patch.Info
+		if patch.Err != nil {
+			allErrs = append(allErrs, fmt.Errorf("error: %s/%s %v\n", info.Mapping.Resource, info.Name, patch.Err))
+			continue
+		}
+
+		//no changes
+		if string(patch.Patch) == "{}" || len(patch.Patch) == 0 {
+			allErrs = append(allErrs, fmt.Errorf("info: %s %q was not changed\n", info.Mapping.Resource, info.Name))
+			continue
+		}
+
+		if o.Local || o.DryRun {
+			if err := cmdutil.PrintObject(o.Cmd, patch.Info.AsVersioned(), o.Out); err != nil {
+				return err
+			}
+			continue
+		}
+
+		obj, err := resource.NewHelper(info.Client, info.Mapping).Patch(info.Namespace, info.Name, types.StrategicMergePatchType, patch.Patch)
+		if err != nil {
+			allErrs = append(allErrs, fmt.Errorf("failed to patch data to : %v\n", err))
+			continue
+		}
+		info.Refresh(obj, true)
+
+		if len(o.Output) > 0 {
+			if err := cmdutil.PrintObject(o.Cmd, info.AsVersioned(), o.Out); err != nil {
+				return err
+			}
+			continue
+		}
+		cmdutil.PrintSuccess(o.ShortOutput, o.Out, info.Object, o.DryRun, "data updated")
+	}
+	return utilerrors.NewAggregate(allErrs)
+}
+
+func getData(fileSources, literalSources []string, envFileSource string) (map[string][]byte, error) {
+
+	data := map[string][]byte{}
+	if len(fileSources) > 0 {
+		if err := handleFromFileSources(data, fileSources); err != nil {
+			return nil, err
+		}
+	}
+	if len(literalSources) > 0 {
+		if err := handleFromLiteralSources(data, literalSources); err != nil {
+			return nil, err
+		}
+	}
+	if len(envFileSource) > 0 {
+		if err := handleFromEnvFileSource(data, envFileSource); err != nil {
+			return nil, err
+		}
+	}
+	
+	return data, nil
+}
+
+//Note: the obj mutates in the function
+func updateDataForObject(obj runtime.Object, data interface{}, fn func(interface{}, interface{}) (bool, interface{})) (bool, error) {
```

Comment:
I don't see the point of the function parameter, since it's the only parameter ever taken. Can you call the addData function directly in here?